# commented out as xlsx export is not fixed yet: diff -rup sc-orig/source/filter/xlsx/xlsx-xcl97rec.cxx sc/source/filter/xlsx/xlsx-xcl97rec.cxx
# commented out as xlsx export is not fixed yet: --- sc-orig/source/filter/xlsx/xlsx-xcl97rec.cxx	2009-07-14 17:32:13.000000000 +0200
# commented out as xlsx export is not fixed yet: +++ sc/source/filter/xlsx/xlsx-xcl97rec.cxx	2009-07-14 17:35:50.000000000 +0200
# commented out as xlsx export is not fixed yet: @@ -1162,7 +1162,7 @@ void XclObjAny::SaveXml( XclExpXmlStream
# commented out as xlsx export is not fixed yet:  
# commented out as xlsx export is not fixed yet:      sax_fastparser::FSHelperPtr pDrawing = rStrm.GetCurrentStream();
# commented out as xlsx export is not fixed yet:  
# commented out as xlsx export is not fixed yet: -    ShapeExport aDML( XML_xdr, pDrawing, &rStrm, DrawingML::DOCUMENT_XLSX );
# commented out as xlsx export is not fixed yet: +    ShapeExport aDML( XML_xdr, pDrawing, NULL, &rStrm, DrawingML::DOCUMENT_XLSX );
# commented out as xlsx export is not fixed yet:  
# commented out as xlsx export is not fixed yet:      pDrawing->startElement( FSNS( XML_xdr, XML_twoCellAnchor ), // OOXTODO: oneCellAnchor, absoluteAnchor
# commented out as xlsx export is not fixed yet:              XML_editAs, GetEditAs( *this ),
diff -rup sd-orig/source/filter/pptx/epptooxml.hxx sd/source/filter/pptx/epptooxml.hxx
--- sd-orig/source/filter/pptx/epptooxml.hxx	2009-07-14 15:44:07.000000000 +0200
+++ sd/source/filter/pptx/epptooxml.hxx	2009-07-14 17:21:34.000000000 +0200
@@ -3,7 +3,7 @@
 
 #include <oox/core/xmlfilterbase.hxx>
 #include <oox/helper/zipstorage.hxx>
-
+#include <oox/export/shapes.hxx>
 #include "epptbase.hxx"
 
 namespace com { namespace sun { namespace star {
@@ -125,6 +125,8 @@ private:
     sal_Bool mbCreateNotes;
 
     static sal_Int32 nStyleLevelToken[5];
+
+    ::oox::drawingml::ShapeExport::ShapeHashMap maShapeMap;
 };
 
 }
diff -rup sd-orig/source/filter/pptx/pptx-epptooxml.cxx sd/source/filter/pptx/pptx-epptooxml.cxx
--- sd-orig/source/filter/pptx/pptx-epptooxml.cxx	2009-07-14 15:44:07.000000000 +0200
+++ sd/source/filter/pptx/pptx-epptooxml.cxx	2009-07-14 17:27:36.000000000 +0200
@@ -79,7 +79,7 @@ class PowerPointShapeExport : public Sha
     PageType            mePageType;
     sal_Bool            mbMaster;
 public:
-                        PowerPointShapeExport( FSHelperPtr pFS, PowerPointExport* pFB );
+    PowerPointShapeExport( FSHelperPtr pFS, ShapeHashMap* pShapeMap, PowerPointExport* pFB );
     void                SetMaster( sal_Bool bMaster );
     void                SetPageType( PageType ePageType );
     ShapeExport&        WriteNonVisualProperties( Reference< XShape > xShape );
@@ -92,8 +92,8 @@ public:
     sal_Bool WritePlaceholder( Reference< XShape > xShape, PlaceholderType ePlaceholder, sal_Bool bMaster );
 };
 
-PowerPointShapeExport::PowerPointShapeExport( FSHelperPtr pFS, PowerPointExport* pFB )
-    : ShapeExport( XML_p, pFS, pFB )
+	PowerPointShapeExport::PowerPointShapeExport( FSHelperPtr pFS, ShapeHashMap* pShapeMap, PowerPointExport* pFB )
+	    : ShapeExport( XML_p, pFS, pShapeMap, pFB )
     , mrExport( *pFB )
 {
 }
@@ -218,6 +218,7 @@ bool PowerPointExport::importDocument()
 bool PowerPointExport::exportDocument() throw()
 {
     DrawingML::ResetCounters();
+    maShapeMap.clear ();
 
     addRelation( US( "http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" ), S( "ppt/presentation.xml" ) );
 
@@ -250,7 +251,9 @@ bool PowerPointExport::exportDocument()
 
     commit();
 
-	return true;
+    maShapeMap.clear ();
+
+    return true;
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------------------------------
@@ -274,7 +277,7 @@ void PowerPointExport::ImplWriteBackgrou
     pFS->startElementNS( XML_p, XML_bg, FSEND );
     pFS->startElementNS( XML_p, XML_bgPr, FSEND );
 
-    PowerPointShapeExport( pFS, this ).WriteFill( rXPropSet );
+    PowerPointShapeExport( pFS, &maShapeMap, this ).WriteFill( rXPropSet );
 
     pFS->endElementNS( XML_p, XML_bgPr );
     pFS->endElementNS( XML_p, XML_bg );
@@ -631,7 +634,7 @@ void PowerPointExport::WriteAnimationTar
     if( rXShape.is() ) {
 	pFS->startElementNS( XML_p, XML_tgtEl, FSEND );
 	pFS->singleElementNS( XML_p, XML_spTgt,
-			      XML_spid, I32S( ShapeExport::GetShapeID( rXShape ) ),
+			      XML_spid, I32S( ShapeExport::GetShapeID( rXShape, &maShapeMap ) ),
 			      FSEND );
 	pFS->endElementNS( XML_p, XML_tgtEl );
     }
@@ -1440,7 +1443,7 @@ void PowerPointExport::ImplWriteLayout(
 
 void PowerPointExport::WriteShapeTree( FSHelperPtr pFS, PageType ePageType, sal_Bool bMaster )
 {
-    PowerPointShapeExport aDML( pFS, this );
+    PowerPointShapeExport aDML( pFS, &maShapeMap, this );
     aDML.SetMaster( bMaster );
     aDML.SetPageType( ePageType );
     sal_uInt32 nShapes;
diff -rup oox-orig/inc/oox/export/shapes.hxx oox/inc/oox/export/shapes.hxx
--- oox-orig/inc/oox/export/shapes.hxx	2009-07-14 15:43:57.000000000 +0200
+++ oox/inc/oox/export/shapes.hxx	2009-07-14 17:27:03.000000000 +0200
@@ -52,16 +52,7 @@ namespace oox { namespace drawingml {
 
 class OOX_DLLPUBLIC ShapeExport : public DrawingML {
 
-protected:
-    sal_Int32           mnShapeIdMax, mnPictureIdMax;
-
 private:
-    sal_Int32           mnXmlNamespace;
-    Fraction            maFraction;
-    MapMode             maMapModeSrc, maMapModeDest;
-
-    ::com::sun::star::awt::Size MapSize( const ::com::sun::star::awt::Size& ) const;
-
     struct ShapeCheck
     {
         bool operator()( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape> s1, const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape> s2 ) const
@@ -77,11 +68,25 @@ private:
         size_t operator()( const ::com::sun::star::uno::Reference < ::com::sun::star::drawing::XShape > ) const;
     };
 
+public:
     typedef std::hash_map< const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape>, sal_Int32, ShapeHash, ShapeCheck> ShapeHashMap;
-    static ShapeHashMap saShapeMap;
+
+protected:
+    sal_Int32           mnShapeIdMax, mnPictureIdMax;
+
+private:
+    sal_Int32           mnXmlNamespace;
+    Fraction            maFraction;
+    MapMode             maMapModeSrc, maMapModeDest;
+
+    ::com::sun::star::awt::Size MapSize( const ::com::sun::star::awt::Size& ) const;
+
+    ShapeHashMap maShapeMap;
+    ShapeHashMap* mpShapeMap;
 
 public:
-    ShapeExport( sal_Int32 nXmlNamespace, ::sax_fastparser::FSHelperPtr pFS, ::oox::core::XmlFilterBase* pFB = NULL, DocumentType eDocumentType = DOCUMENT_PPTX );
+
+    ShapeExport( sal_Int32 nXmlNamespace, ::sax_fastparser::FSHelperPtr pFS, ShapeHashMap* pShapeMap = NULL, ::oox::core::XmlFilterBase* pFB = NULL, DocumentType eDocumentType = DOCUMENT_PPTX );
     virtual ~ShapeExport() {}
 
     sal_Int32           GetXmlNamespace() const;
@@ -160,8 +165,9 @@ public:
                         WriteUnknownShape( ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > xShape );
 
     sal_Int32 GetNewShapeID( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > rShape );
-    static sal_Int32 GetNewShapeID( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > rShape, ::oox::core::XmlFilterBase* pFB );
-    static sal_Int32 GetShapeID( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > rShape );
+    sal_Int32 GetNewShapeID( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > rShape, ::oox::core::XmlFilterBase* pFB );
+    sal_Int32 GetShapeID( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > rShape );
+    static sal_Int32 GetShapeID( const ::com::sun::star::uno::Reference< ::com::sun::star::drawing::XShape > rShape, ShapeHashMap* pShapeMap );
 };
 
 }}
diff -rup oox-orig/source/export/shapes.cxx oox/source/export/shapes.cxx
--- oox-orig/source/export/shapes.cxx	2009-07-14 15:44:00.000000000 +0200
+++ oox/source/export/shapes.cxx	2009-07-14 17:26:27.000000000 +0200
@@ -361,7 +361,7 @@ namespace oox { namespace drawingml {
     if ( GETA(propName) ) \
         mAny >>= variable;
 
-ShapeExport::ShapeExport( sal_Int32 nXmlNamespace, FSHelperPtr pFS, XmlFilterBase* pFB, DocumentType eDocumentType )
+ShapeExport::ShapeExport( sal_Int32 nXmlNamespace, FSHelperPtr pFS, ShapeHashMap* pShapeMap, XmlFilterBase* pFB, DocumentType eDocumentType )
     : DrawingML( pFS, pFB, eDocumentType )
     , mnShapeIdMax( 1 )
     , mnPictureIdMax( 1 )
@@ -369,6 +369,7 @@ ShapeExport::ShapeExport( sal_Int32 nXml
     , maFraction( 1, 576 )
     , maMapModeSrc( MAP_100TH_MM )
     , maMapModeDest( MAP_INCH, Point(), maFraction, maFraction )
+    , mpShapeMap( pShapeMap ? pShapeMap : &maShapeMap )
 {
 }
 
@@ -989,24 +990,27 @@ sal_Int32 ShapeExport::GetNewShapeID( co
 
     sal_Int32 nID = pFB->GetUniqueId();
 
-    saShapeMap[ rXShape ] = nID;
+    (*mpShapeMap)[ rXShape ] = nID;
 
     return nID;
 }
 
 sal_Int32 ShapeExport::GetShapeID( const Reference< XShape > rXShape )
 {
+    return GetShapeID( rXShape, mpShapeMap );
+}
+
+sal_Int32 ShapeExport::GetShapeID( const Reference< XShape > rXShape, ShapeHashMap* pShapeMap )
+{
     if( !rXShape.is() )
         return -1;
 
-    ShapeHashMap::const_iterator aIter = saShapeMap.find( rXShape );
+    ShapeHashMap::const_iterator aIter = pShapeMap->find( rXShape );
 
-    if( aIter == saShapeMap.end() )
+    if( aIter == pShapeMap->end() )
         return -1;
 
     return aIter->second;
 }
 
-ShapeExport::ShapeHashMap ShapeExport::saShapeMap;
-
 } }
